sphinx/util/inspect.py

```diff
@@ -557,6 +557,45 @@ def signature_from_str(signature: str) -> inspect.Signature:
     return inspect.Signature(params, return_annotation=return_annotation)
 
 
+def signature_from_ast(node: ast.FunctionDef) -> inspect.Signature:
+    """Return a Signature object for the given AST *node*."""
+    params = []
+    if hasattr(node.args, "posonlyargs"):  # for py38+
+        for arg in node.args.posonlyargs:  # type: ignore
+            param = Parameter(arg.arg, Parameter.POSITIONAL_ONLY,
+                              annotation=ast_unparse(arg.annotation))
+            params.append(param)
+
+    for i, arg in enumerate(node.args.args):
+        if len(node.args.args) - i <= len(node.args.defaults):
+            default = ast_unparse(node.args.defaults[-len(node.args.args) + i])
+        else:
+            default = Parameter.empty
```

Comment:
Possibly simpler as:
```suggestion
    defaults = [Parameter.empty] * (len(node.args.args) - len(node.args.defaults)) + node.args.defaults
    for arg, default in enumerate(node.args.args, defaults):
```